src/tr2/game/objects/general/earthquake.c

```diff
@@ -1,7 +1,51 @@
 #include "game/objects/general/earthquake.h"
 
+#include "game/items.h"
 #include "game/objects/common.h"
-#include "global/funcs.h"
+#include "game/random.h"
+#include "game/room.h"
+#include "game/sound.h"
+#include "global/vars.h"
+
+static void M_Activate(int16_t earth_item_num);
+
+static void M_Activate(const int16_t earth_item_num)
+{
+    ITEM *const earth_item = Item_Get(earth_item_num);
+    Item_AddActive(earth_item_num);
+    earth_item->status = IS_ACTIVE;
+    earth_item->flags = IF_CODE_BITS;
+    earth_item->timer = 0;
+}
+
+void __cdecl Earthquake_Control(const int16_t item_num)
+{
+    const ITEM *const item = Item_Get(item_num);
+    if (Random_GetDraw() < 512) {
+        Sound_Effect(SFX_DRAGON_FEET, NULL, SPM_NORMAL);
+        g_Camera.bounce = -200;
+    }
+
+    GAME_OBJECT_ID object_id_to_activate;
+    const int32_t random = Random_GetControl();
+    if (random < 512) {
+        object_id_to_activate = O_FLAME_EMITTER;
+    } else if (random < 1024) {
+        object_id_to_activate = O_FALLING_CEILING;
+    } else {
+        return;
+    }
+
+    int16_t earth_item_num = Room_Get(item->room_num)->item_num;
+    while (earth_item_num != NO_ITEM) {
+        const ITEM *const earth_item = Item_Get(earth_item_num);
+        if (earth_item->object_id == object_id_to_activate
+            && item->status != IS_ACTIVE && item->status == IS_DEACTIVATED) {
+            M_Activate(earth_item_num);
+            break;
+        }
```

Comment:
I guess this is where Dragon's Lair is getting stuck, do we need to traverse the item linked list? Also note sure if `item->status != IS_ACTIVE && item->status == IS_DEACTIVATED` is correct - should it just be one status?

Reply:
Thanks – addressed. And yes the status checks were quite botched, they were supposed to both test for inequality and on `earth_item` rather than `item` (which is the earthquake emitter).